crates/hulk/src/nao/hula.rs

```diff
@@ -377,10 +380,30 @@ pub struct ControlStorage {
     pub stiffness: JointsArray,
 }
 
-pub fn read_from_hula(stream: &mut UnixStream) -> Result<StateStorage> {
-    let mut read_buffer = [0; size_of::<StateStorage>()];
-    stream.read_exact(&mut read_buffer)?;
-    Ok(unsafe { read(read_buffer.as_ptr() as *const StateStorage) })
+pub fn read_from_hula(
+    stream: &mut UnixStream,
+    reader: &mut DoubleBufferedReader,
+) -> Result<StateStorage> {
+    reader
+        .read(stream, |file_descriptor| unsafe {
+            let mut set = MaybeUninit::<fd_set>::uninit();
+            FD_ZERO(set.as_mut_ptr());
+            let mut set = set.assume_init();
+            FD_SET(file_descriptor, &mut set);
+            if select(
```

Comment:
Can you elaborate what this does? Is there nothing that already implements such a behavior as an API function on some file descriptor type?

Reply:
Since this is Unix only functionality and Rust's standard library does not provide such high-level constructs, there is nothing available officially. I decided to use the libc call directly which is unsafe but has is a well-known API and does not introduce heavy abstractions. My goal was to find a thing that allows me to simply wait for a single file descriptor becoming readable. The other Linux polling methods would also work but `select()` is one of the simpler ways to do it. We could search for a safe abstraction over `select()` to get rid of the `unsafe {}` here. What is your opinion or do you know another way to use standard functionality that I didn't see?

Reply:
No, I just assumed, there already is a safe abstraction. I now, that for example https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html#method.readable does something like you did manually. Maybe there is a crate for these functionalities out there?!

Reply:
Maybe this? https://github.com/DoumanAsh/selecting Sadly there is no `readable()` in non-async land as far as I know. IMHO the benefit for such simple functionality is debatable.